### DISCUSSION TITLE
Tasmota32 reports different switch state on MQTT and on tasmota.get_switches()

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#19052] [General]
I noticed, that on my ESP32 board I see different states of switches:
```
tasmota.get_switches()
[false, false, false, false, false, false]

14:19:17.936 MQT: tele/small_tasmota32/SENSOR = {"Time":"2023-07-06T14:19:17","Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"ON","Switch5":"ON","Switch6":"ON"}

**Comment by sfromis** [comment]
That's not different states, merely different ways of representing high/low.

It is as correct (or incorrect) to represent high as false/ON, or represent low as true/OFF.

For simple cases like a switch input connecting to ground when pushed, what you get in Berry is more intuitive, while it sounds "wrong" to say that high is false, and low is true.

**Reply by msetina** [reply]
OK. Is Switch1 set differently on Tasmota and Tasmota32? 
For Tasmota it says it uses internal pullup resistor. This result is like it does not. I had two more GPIOs which are on the board as inputs and not outputs with no internal pullups on Switch7 and Switch8 and those were OFF in MQTT.
This was my first confusion, then I saw the mentioned state and did not know where the problem is.

**Reply by sfromis** [reply]
What exactly do you mean by "differently", in what specific circumstances and expectations?

Tasmota should by default enable the weak internal pullup on both ESP8266 and ESP32. If you configure without internal pullup (`Switch_n`), results are unpredictable unless you make sure to have external pullup. 

What I see on ESP32 is that regular switches with the default internal pullup, and not pushed, appear as `"Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"ON"` in the /SENSOR MQTT payloads, and as `[false, false, false, false]` when using `tasmota.get_switches()`. This is exactly as expected, and matches what you posted first. Nothing unusual in that, as long as you do not make assumptions about a specific "correct" meaning of high/low, on/off and true/false.

**Reply by msetina** [reply]
This is my confusion. I have a fixed idea HIGH/1/ON/true  and on the other side LOW/0/OFF/false.
I see there is inverted mode on a switch.

**Reply by sfromis** [reply]
Inverted mode on a Switch is not on the input, but by selecting an inverted SwitchMode.

**Reply by msetina** [reply]
Yes. This is how I understand inverted mode.
What provides confusion is that MQTT reports switch as ON in a case where there is no switch connected to the GPIO. The Berry call returns false, which shows that the switch in off/not closed, which represents reality. With inverted mode MQTT Switch state is reported as OFF which is in line with false value reported by Berry call.
As I understand the GPIO logic for buttons and swithes in inverted as GPIO needs to be connected to GND to indicate closed/ON state, because we use pullup resistor.
For my case of GPIO ports without pullup I know I need external pullup, I just did not check the documentation of the board.

**Reply by sfromis** [reply]
Yes, standard MCU logic is that such inputs connect a pin to ground when activated, and internal pullup to provide a defined state when not activated. This means that no voltage is applied to the switch/button input, as the ground connection when activated will draw away the very small pullup current.

What do you mean by "GPIO ports without pullup"? Tasmota will **always** enable internal pullup, unless you specifically config it to not do that. With normal short wires, you need no external pullup, unless you have "more challenging" circumstances like electrical noise, or longer wires.

My point about mentioning inverted mode on a Switch pin is that it is quite different from a Button pin, where you have inverted or not as part of the gpio config, instead of something like a SwitchMode.

**Reply by msetina** [reply]
GPIO ports without pullup on my board are GPIO port that are marked as input only and as per documentation do not have internal pullup resistor. I my case they are GPIO34 and GPIO35.
